synapse/metrics/_reactor_metrics.py

```diff
@@ -34,52 +62,94 @@
+            return self._wrapped_method
+
+        return getattr(self._wrapped, item)
 
 
 class ReactorLastSeenMetric(Collector):
-    def __init__(self, epoll_wrapper: EpollWrapper):
-        self._epoll_wrapper = epoll_wrapper
+    def __init__(self, call_wrapper: CallWrapper):
+        self._call_wrapper = call_wrapper
 
     def collect(self) -> Iterable[Metric]:
         cm = GaugeMetricFamily(
             "python_twisted_reactor_last_seen",
             "Seconds since the Twisted reactor was last seen",
         )
-        cm.add_metric([], time.time() - self._epoll_wrapper.last_polled)
+        cm.add_metric([], time.time() - self._call_wrapper.last_polled)
         yield cm
 
 
-try:
-    # if the reactor has a `_poller` attribute, which is an `epoll` object
-    # (ie, it's an EPollReactor), we wrap the `epoll` with a thing that will
-    # measure the time between ticks
-    from select import epoll  # type: ignore[attr-defined]
-
-    poller = reactor._poller  # type: ignore[attr-defined]
-except (AttributeError, ImportError):
-    pass
-else:
-    if isinstance(poller, epoll):
-        poller = EpollWrapper(poller)
-        reactor._poller = poller  # type: ignore[attr-defined]
-        REGISTRY.register(ReactorLastSeenMetric(poller))
+# Twisted has already select a reasonable reactor for us, so assumptions can be
+# made about the shape.
+wrapper = None
+if isinstance(reactor, (PollReactor, EPollReactor)):
+    reactor._poller = ObjWrapper(reactor._poller, "poll")  # type: ignore[attr-defined]
+    wrapper = reactor._poller._wrapped_method  # type: ignore[attr-defined]
+
+elif isinstance(reactor, selectreactor.SelectReactor):
+    # Twisted uses a module-level _select function.
+    wrapper = selectreactor._select = CallWrapper(selectreactor._select)
+
+elif isinstance(reactor, AsyncioSelectorReactor):
+    # For asyncio look at the underlying asyncio event loop.
+    asyncio_loop = reactor._asyncioEventloop  # A sub-class of BaseEventLoop,
+
+    # If an unexpected asyncio loop implementation is used, these might fail.
+    try:
+        # A sub-class of BaseSelector.
+        selector = asyncio_loop._selector  # type: ignore[attr-defined]
+
+        if isinstance(selector, SelectSelector):
+            wrapper = selector._select = CallWrapper(selector._select)  # type: ignore[attr-defined]
+
+        # poll, epoll, and /dev/poll.
+        elif isinstance(selector, _PollLikeSelector):
+            selector._selector = ObjWrapper(selector._selector, "poll")  # type: ignore[attr-defined]
+            wrapper = selector._selector._wrapped_method  # type: ignore[attr-defined]
+
+        elif isinstance(selector, KqueueSelector):
+            selector._selector = ObjWrapper(selector._selector, "control")  # type: ignore[attr-defined]
+            wrapper = selector._selector._wrapped_method  # type: ignore[attr-defined]
+
+        # XXX Will not work on (Windows-only) ProactorEventLoop.
+    except AttributeError:
```

Comment:
Maybe this should be wider and have this entire block covered?

Reply:
Yes, I think we should have a very broad `try/except` here, that also includes access to underscored variables. I'd also be tempted to make this an `except Exception`